File: web/src/source/MockSource.ts
/**
 * Synthetic data source (§8): drives realistic packets on timers with no hardware, so
 * the entire app — dashboard, recording, calibration, gating, export — is demoable.
 * Emits typed packets with already-monotonic device timestamps.
 */
import type { BiozPacket, EcgPacket, MetricsPacket, PpgPacket } from '../ble/packets';
import { DEFAULT_K, DEFAULT_RATE_HZ, StreamBit, type StreamKind } from '../ble/protocol';
import { ContactMotionEstimator } from '../dsp/bioz';
import { US_PER_S } from '../dsp/constants';
import { BaseDataSource, type ConnectionState } from './DataSource';
import { MockSignal, type MockParams } from './mockSignal';

const TICK_MS = 40;
const METRICS_MS = 500;

interface StreamState {
  rateHz: number;
  k: number;
  nextIndex: number;
  seq: number;
}

export class MockSource extends BaseDataSource {
  readonly isMock = true;
  clockOffsetUs: number | null = null;
  deviceName = 'VitalView Mock';

  private signal: MockSignal;
  private streams: Record<StreamKind, StreamState>;
  private enabledMask = 0;
  private running = false;
  private startPerfMs = 0;
  private batteryPct = 100;
  private timer: ReturnType<typeof setInterval> | null = null;
  private metricsTimer: ReturnType<typeof setInterval> | null = null;
  private statusTimer: ReturnType<typeof setInterval> | null = null;
  private lastMetricsDeviceUs = 0;
  private readonly contact: ContactMotionEstimator;

  constructor(params: Partial<MockParams> = {}) {
    super();
    this.signal = new MockSignal(params);
    this.contact = new ContactMotionEstimator(this.signal.p.biozRateHz);
    this.streams = {
      ecg: { rateHz: DEFAULT_RATE_HZ.ecg, k: DEFAULT_K.ecg, nextIndex: 0, seq: 0 },
      bioz: { rateHz: DEFAULT_RATE_HZ.bioz, k: DEFAULT_K.bioz, nextIndex: 0, seq: 0 },
      ppg: { rateHz: DEFAULT_RATE_HZ.ppg, k: DEFAULT_K.ppg, nextIndex: 0, seq: 0 },
    };
  }

  /** Live-adjust the synthetic signal (HR/PAT/SpO2/etc.) from Settings. */
  setParams(params: Partial<MockParams>): void {
    Object.assign(this.signal.p, params);
  }

  getParams(): MockParams {
    return this.signal.p;
  }

  /** Inject a motion burst now (drives the motion/gating demo). */
  injectMotion(durationMs = 2000, severity = 4): void {
    const now = this.deviceNowUs();
    this.signal.p.motionBursts.push({
      startUs: now,
      endUs: now + durationMs * 1000,
      severity,
    });
  }

  async connect(): Promise<void> {
    this.setState('connecting');
    await delay(150);
    this.setState('connected');
    this.startPerfMs = now();
    this.getInfoSync();
    this.statusTimer = setInterval(() => this.emitStatus(), 1000);
  }

  async disconnect(): Promise<void> {
    await this.stop();
    clearTimer(this.statusTimer);
    this.statusTimer = null;
    this.setState('disconnected');
  }

  async start(streamMask: number): Promise<void> {
    this.enabledMask = streamMask;
    this.startPerfMs = now();
    this.lastMetricsDeviceUs = 0;
    for (const s of Object.values(this.streams)) {
      s.nextIndex = 0;
      s.seq = 0;
    }
    this.contact.reset();
    this.running = true;
    this.timer = setInterval(() => this.tick(), TICK_MS);
    this.metricsTimer = setInterval(() => this.emitMetrics(), METRICS_MS);
  }

  async stop(): Promise<void> {
    this.running = false;
    clearTimer(this.timer);
    clearTimer(this.metricsTimer);
    this.timer = null;
    this.metricsTimer = null;
  }

  async setRate(stream: StreamKind, hz: number): Promise<void> {
    const s = this.streams[stream];
    s.rateHz = hz;
    if (this.running) s.nextIndex = Math.floor(this.elapsedS() * hz);
    if (stream === 'bioz') this.signal.p.biozRateHz = hz;
    if (stream === 'ecg') this.signal.p.ecgRateHz = hz;
    if (stream === 'ppg') this.signal.p.ppgRateHz = hz;
  }

  async syncClock(): Promise<void> {
    await delay(20);
    const deviceTus = this.deviceNowUs();
    this.clockOffsetUs = Date.now() * 1000 - deviceTus;
  }

  async getInfo(): Promise<void> {
    this.getInfoSync();
  }

  // ---- internals ----

  private getInfoSync(): void {
    this.emitter.emit('info', {
      kind: 'info',
      firmwareVersion: '1.0.0-mock',
      deviceId: 'mock00000001',
      capabilities: StreamBit.ecg | StreamBit.bioz | StreamBit.ppg,
    });
    this.emitter.emit('battery', this.batteryPct);
  }

  private elapsedS(): number {
    return (now() - this.startPerfMs) / 1000;
  }

  private deviceNowUs(): number {
    return Math.round(this.elapsedS() * US_PER_S);
  }

  private tick(): void {
    if (!this.running) return;
    const elapsed = this.elapsedS();
    if (this.enabledMask & StreamBit.ecg) this.tickEcg(elapsed);
    if (this.enabledMask & StreamBit.bioz) this.tickBioz(elapsed);
    if (this.enabledMask & StreamBit.ppg) this.tickPpg(elapsed);
  }

  private tickEcg(elapsedS: number): void {
    const s = this.streams.ecg;
    const target = Math.floor(elapsedS * s.rateHz);
    while (s.nextIndex < target) {
      const count = Math.min(s.k, target - s.nextIndex);
      const first = s.nextIndex;
      const tUs = Math.round((first * US_PER_S) / s.rateHz);
      const samples = new Int32Array(count);
      for (let i = 0; i < count; i++) {
        const t = Math.round(((first + i) * US_PER_S) / s.rateHz);
        samples[i] = Math.round(this.signal.ecgAt(t));
      }
      const p: EcgPacket = { kind: 'ecg', seq: s.seq, tUs, sampleRateHz: s.rateHz, samples };
      this.emitter.emit('ecg', p);
      s.nextIndex += count;
      s.seq = (s.seq + 1) & 0xffff;
    }
  }

  private tickBioz(elapsedS: number): void {
    const s = this.streams.bioz;
    const target = Math.floor(elapsedS * s.rateHz);
    while (s.nextIndex < target) {
      const count = Math.min(s.k, target - s.nextIndex);
      const first = s.nextIndex;
      const tUs = Math.round((first * US_PER_S) / s.rateHz);
      const z0 = Math.round(this.signal.z0At(tUs));
      const dz = new Int32Array(count);
      for (let i = 0; i < count; i++) {
        const t = Math.round(((first + i) * US_PER_S) / s.rateHz);
        dz[i] = Math.round(this.signal.dzAt(t));
        this.contact.pushDz(dz[i]);
      }
      this.contact.pushZ0(z0);
      const p: BiozPacket = { kind: 'bioz', seq: s.seq, tUs, sampleRateHz: s.rateHz, z0Milliohm: z0, dz };
      this.emitter.emit('bioz', p);
      s.nextIndex += count;
      s.seq = (s.seq + 1) & 0xffff;
    }
  }

  private tickPpg(elapsedS: number): void {
    const s = this.streams.ppg;
    const target = Math.floor(elapsedS * s.rateHz);
    while (s.nextIndex < target) {
      const count = Math.min(s.k, target - s.nextIndex);
      const first = s.nextIndex;
      const tUs = Math.round((first * US_PER_S) / s.rateHz);
      const green = new Uint32Array(count);
      const red = new Uint32Array(count);
      const ir = new Uint32Array(count);
      for (let i = 0; i < count; i++) {
        const t = Math.round(((first + i) * US_PER_S) / s.rateHz);
        green[i] = Math.max(0, Math.round(this.signal.greenAt(t)));
        red[i] = Math.max(0, Math.round(this.signal.redAt(t)));
        ir[i] = Math.max(0, Math.round(this.signal.irAt(t)));
      }
      const p: PpgPacket = { kind: 'ppg', seq: s.seq, tUs, sampleRateHz: s.rateHz, green, red, ir };
      this.emitter.emit('ppg', p);
      s.nextIndex += count;
      s.seq = (s.seq + 1) & 0xffff;
    }
  }

  private emitMetrics(): void {
    if (!this.running) return;
    const tUs = this.deviceNowUs();
    const beats = this.signal.rPeakTimes(this.lastMetricsDeviceUs, tUs);
    this.lastMetricsDeviceUs = tUs;
    const p: MetricsPacket = {
      kind: 'metrics',
      tUs,
      hrBpm: this.signal.p.hrBpm,
      spo2Pct: this.signal.p.spo2Target,
      contactQuality: this.contact.contactQuality(),
      motion: this.contact.motion(),
      patUs: this.signal.p.patMs * 1000,
      sbpMmHg: null, // firmware reports BP uncalibrated; calibration is per-subject in-app
      dbpMmHg: null,
      rpeak: beats.length > 0,
    };
    this.emitter.emit('metrics', p);
  }

  private emitStatus(): void {
    // Slow battery drain for realism.
    this.batteryPct = Math.max(0, this.batteryPct - 0.05);
    this.emitter.emit('battery', Math.round(this.batteryPct));
    this.emitter.emit('status', {
      kind: 'status',
      streamingMask: this.running ? this.enabledMask : 0,
      ecgRateCode: 1,
      biozRateCode: 1,
      ppgRateCode: 1,
      tUs: this.deviceNowUs(),
      batteryPct: Math.round(this.batteryPct),
      errorFlags: 0,
    });
  }
}

function now(): number {
  return typeof performance !== 'undefined' ? performance.now() : Date.now();
}
function delay(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}
function clearTimer(t: ReturnType<typeof setInterval> | null): void {
  if (t != null) clearInterval(t);
}

export type { ConnectionState };

File: ios/MoniVitals/Source/MockSource.swift
//
//  MockSource.swift
//  MoniVitals
//
//  Synthetic data source: drives realistic packets on timers with no hardware, so the
//  entire app — dashboard, recording, calibration, gating, export — is demoable. Emits
//  typed packets with already-monotonic device timestamps.
//
//  Ported from web/src/source/MockSource.ts. Uses GCD timers on a serial queue; elapsed
//  time uses a monotonic clock (systemUptime) and the clock offset uses wall time, to
//  mirror the web's performance.now()/Date.now() split.
//

import Foundation

private let TICK_MS = 40.0
private let METRICS_MS = 500.0

private final class StreamState {
    var rateHz: Int
    var k: Int
    var nextIndex: Int = 0
    var seq: Int = 0
    init(rateHz: Int, k: Int) {
        self.rateHz = rateHz
        self.k = k
    }
}

final class MockSource: BaseDataSource {
    override var isMock: Bool { true }
    private var _clockOffsetUs: Double? = nil
    override var clockOffsetUs: Double? { _clockOffsetUs }
    private var _deviceName: String? = "MoniVitals Mock"
    override var deviceName: String? { _deviceName }

    private var signal: MockSignal
    private var streams: [StreamKind: StreamState]
    private var enabledMask: UInt8 = 0
    private var running = false
    private var startUptime: TimeInterval = 0
    private var batteryPct = 100.0
    private var tickTimer: DispatchSourceTimer?
    private var metricsTimer: DispatchSourceTimer?
    private var statusTimer: DispatchSourceTimer?
    private var lastMetricsDeviceUs = 0.0
    private let contact: ContactMotionEstimator
    private let queue = DispatchQueue(label: "com.monivitals.mocksource")

    init(_ params: MockParams = .default) {
        self.signal = MockSignal(params)
        self.contact = ContactMotionEstimator(fsBioz: Double(params.biozRateHz))
        self.streams = [
            .ecg: StreamState(rateHz: BLEProtocol.defaultRateHz(.ecg), k: BLEProtocol.defaultK(.ecg)),
            .bioz: StreamState(rateHz: BLEProtocol.defaultRateHz(.bioz), k: BLEProtocol.defaultK(.bioz)),
            .ppg: StreamState(rateHz: BLEProtocol.defaultRateHz(.ppg), k: BLEProtocol.defaultK(.ppg)),
        ]
        super.init()
    }

    /// Live-adjust the synthetic signal (HR/PAT/SpO2/etc.) from Settings.
    /// `signal.p` is read by the emission timers on `queue`, so mutate it there.
    func setParams(_ mutate: @escaping (inout MockParams) -> Void) {
        queue.async { [weak self] in
            guard let self else { return }
            mutate(&self.signal.p)
        }
    }

    func getParams() -> MockParams { queue.sync { signal.p } }

    /// Inject a motion burst now (drives the motion/gating demo).
    func injectMotion(durationMs: Double = 2000, severity: Double = 4) {
        queue.async { [weak self] in
            guard let self else { return }
            let now = self.deviceNowUs()
            self.signal.p.motionBursts.append(
                MotionBurst(startUs: now, endUs: now + durationMs * 1000, severity: severity))
        }
    }

    override func connect() {
        setState(.connecting)
        queue.asyncAfter(deadline: .now() + 0.15) { [weak self] in
            guard let self else { return }
            self.setState(.connected)
            self.startUptime = ProcessInfo.processInfo.systemUptime
            self.emitInfo()
            self.statusTimer = self.makeTimer(intervalMs: 1000) { [weak self] in self?.emitStatus() }
        }
    }

    override func disconnect() {
        stop()
        statusTimer?.cancel()
        statusTimer = nil
        setState(.disconnected)
    }

    override func start(streamMask: UInt8) {
        queue.async { [weak self] in
            guard let self else { return }
            self.enabledMask = streamMask
            self.startUptime = ProcessInfo.processInfo.systemUptime
            self.lastMetricsDeviceUs = 0
            for s in self.streams.values {
                s.nextIndex = 0
                s.seq = 0
            }
            self.contact.reset()
            self.running = true
            self.tickTimer = self.makeTimer(intervalMs: TICK_MS) { [weak self] in self?.tick() }
            self.metricsTimer = self.makeTimer(intervalMs: METRICS_MS) { [weak self] in self?.emitMetrics() }
        }
    }

    override func stop() {
        queue.async { [weak self] in
            guard let self else { return }
            self.running = false
            self.tickTimer?.cancel()
            self.metricsTimer?.cancel()
            self.tickTimer = nil
            self.metricsTimer = nil
        }
    }

    override func setRate(_ stream: StreamKind, hz: Int) {
        queue.async { [weak self] in
            guard let self, let s = self.streams[stream] else { return }
            s.rateHz = hz
            if self.running { s.nextIndex = Int(self.elapsedS() * Double(hz)) }
            switch stream {
            case .bioz: self.signal.p.biozRateHz = hz
            case .ecg: self.signal.p.ecgRateHz = hz
            case .ppg: self.signal.p.ppgRateHz = hz
            }
        }
    }

    override func syncClock() {
        queue.asyncAfter(deadline: .now() + 0.02) { [weak self] in
            guard let self else { return }
            let deviceTus = self.deviceNowUs()
            self._clockOffsetUs = Date().timeIntervalSince1970 * 1_000_000 - deviceTus
        }
    }

    override func getInfo() {
        queue.async { [weak self] in self?.emitInfo() }
    }

    // MARK: - internals

    private func makeTimer(intervalMs: Double, _ work: @escaping () -> Void) -> DispatchSourceTimer {
        let t = DispatchSource.makeTimerSource(queue: queue)
        t.schedule(deadline: .now() + .milliseconds(Int(intervalMs)),
                   repeating: .milliseconds(Int(intervalMs)))
        t.setEventHandler(handler: work)
        t.resume()
        return t
    }

    private func emitInfo() {
        events.info.send(InfoMessage(
            firmwareVersion: "1.0.0-mock",
            deviceId: "mock00000001",
            capabilities: Int(BLEProtocol.streamBit(.ecg) | BLEProtocol.streamBit(.bioz) | BLEProtocol.streamBit(.ppg))
        ))
        events.battery.send(Int(batteryPct))
    }

    private func elapsedS() -> Double {
        ProcessInfo.processInfo.systemUptime - startUptime
    }

    private func deviceNowUs() -> Double {
        (elapsedS() * DSPConstants.usPerS).rounded()
    }

    private func tick() {
        if !running { return }
        let elapsed = elapsedS()
        if enabledMask & BLEProtocol.streamBit(.ecg) != 0 { tickEcg(elapsed) }
        if enabledMask & BLEProtocol.streamBit(.bioz) != 0 { tickBioz(elapsed) }
        if enabledMask & BLEProtocol.streamBit(.ppg) != 0 { tickPpg(elapsed) }
    }

    private func tickEcg(_ elapsedS: Double) {
        guard let s = streams[.ecg] else { return }
        let target = Int(elapsedS * Double(s.rateHz))
        while s.nextIndex < target {
            let count = min(s.k, target - s.nextIndex)
            let first = s.nextIndex
            let tUs = (Double(first) * DSPConstants.usPerS / Double(s.rateHz)).rounded()
            var samples = [Int32](repeating: 0, count: count)
            for i in 0..<count {
                let t = (Double(first + i) * DSPConstants.usPerS / Double(s.rateHz)).rounded()
                samples[i] = Int32(signal.ecgAt(t).rounded())
            }
            events.ecg.send(EcgPacket(seq: s.seq, tUs: tUs, sampleRateHz: s.rateHz, samples: samples))
            s.nextIndex += count
            s.seq = (s.seq + 1) & 0xffff
        }
    }

    private func tickBioz(_ elapsedS: Double) {
        guard let s = streams[.bioz] else { return }
        let target = Int(elapsedS * Double(s.rateHz))
        while s.nextIndex < target {
            let count = min(s.k, target - s.nextIndex)
            let first = s.nextIndex
            let tUs = (Double(first) * DSPConstants.usPerS / Double(s.rateHz)).rounded()
            let z0 = Int32(signal.z0At(tUs).rounded())
            var dz = [Int32](repeating: 0, count: count)
            for i in 0..<count {
                let t = (Double(first + i) * DSPConstants.usPerS / Double(s.rateHz)).rounded()
                dz[i] = Int32(signal.dzAt(t).rounded())
                contact.pushDz(Double(dz[i]))
            }
            contact.pushZ0(Double(z0))
            events.bioz.send(BiozPacket(seq: s.seq, tUs: tUs, sampleRateHz: s.rateHz, z0Milliohm: z0, dz: dz))
            s.nextIndex += count
            s.seq = (s.seq + 1) & 0xffff
        }
    }

    private func tickPpg(_ elapsedS: Double) {
        guard let s = streams[.ppg] else { return }
        let target = Int(elapsedS * Double(s.rateHz))
        while s.nextIndex < target {
            let count = min(s.k, target - s.nextIndex)
            let first = s.nextIndex
            let tUs = (Double(first) * DSPConstants.usPerS / Double(s.rateHz)).rounded()
            var green = [UInt32](repeating: 0, count: count)
            var red = [UInt32](repeating: 0, count: count)
            var ir = [UInt32](repeating: 0, count: count)
            for i in 0..<count {
                let t = (Double(first + i) * DSPConstants.usPerS / Double(s.rateHz)).rounded()
                green[i] = UInt32(max(0, signal.greenAt(t).rounded()))
                red[i] = UInt32(max(0, signal.redAt(t).rounded()))
                ir[i] = UInt32(max(0, signal.irAt(t).rounded()))
            }
            events.ppg.send(PpgPacket(seq: s.seq, tUs: tUs, sampleRateHz: s.rateHz, green: green, red: red, ir: ir))
            s.nextIndex += count
            s.seq = (s.seq + 1) & 0xffff
        }
    }

    private func emitMetrics() {
        if !running { return }
        let tUs = deviceNowUs()
        let beats = signal.rPeakTimes(lastMetricsDeviceUs, tUs)
        lastMetricsDeviceUs = tUs
        events.metrics.send(MetricsPacket(
            tUs: tUs,
            hrBpm: signal.p.hrBpm,
            spo2Pct: signal.p.spo2Target,
            contactQuality: contact.contactQuality(),
            motion: contact.motion(),
            patUs: signal.p.patMs * 1000,
            sbpMmHg: nil, // firmware reports BP uncalibrated; calibration is per-subject in-app
            dbpMmHg: nil,
            rpeak: !beats.isEmpty
        ))
    }

    private func emitStatus() {
        // Slow battery drain for realism.
        batteryPct = max(0, batteryPct - 0.05)
        events.battery.send(Int(batteryPct.rounded()))
        events.status.send(StatusMessage(
            streamingMask: running ? enabledMask : 0,
            ecgRateCode: 1,
            biozRateCode: 1,
            ppgRateCode: 1,
            tUs: deviceNowUs(),
            batteryPct: Int(batteryPct.rounded()),
            errorFlags: 0
        ))
    }
}
